### ISSUE TITLE
CSV Import

### ISSUE THREAD

**Issue by sauloperez** [Platoniq/decidim-verifications-direct_verifications#3]
At https://github.com/coopdevs/decidim-coopcat we have use cases with hundreds or even thousands of users which makes pasting data into the text area rather impractical.

Besides, based on https://github.com/Platoniq/decidim-verifications-direct_verifications/pull/2 at decidim-coopcat we also rely on a specific format for the CSV so giving the option to upload the file directly will make things less error-prone.

Finally, according to @microstudi sending a long list of users can cause performance issues. Switching to file upload and making it async should also mitigate that.

**Comment by sauloperez** [comment]
This is a very much-needed UX improvement which we're ready to work on @microstudi .

The most common error we're seeing when importing the census is that users mess up with the format even they can copy/paste an example from https://guia.cercles.coop/index-de-continguts-de-cercles.coop/posada-en-marxa#importacio-del-cens-virtual-de-socis. We better tackle this issue than going down the path of defensive coding IMO.

I'd be happy to share a high-level architecture in the upcoming days.

**Comment by microstudi** [comment]
So glad you're doing this.
I don't know how many hour you want to dedicate to this but it would be nice to have this:
1. upload csv (no specific format required)
2. after download, preview the firs 10 rows and let the user choose (with selects) which column needs to be maped to which imported field.
3. do the import according the configuration

**Comment by sauloperez** [comment]
oops, that involves more effort than what I expected :sweat_smile: . I'm afraid that not having a specific format will explode the complexity of the solution.

What I had in mind was to add a link to a CSV template that users can fill up and upload. That template could have comments other than the header row so users understand the expected format.

I guess a preview will be important in terms of UX but I'd keep it as simple as possible.

**Comment by microstudi** [comment]
all right, we will make improvement later on then

**Comment by sauloperez** [comment]
I think it's the most agile and best approach

**Comment by sauloperez** [comment]
Alright, I think I have a plan. Let's see what you think @microstudi .

Basically, we need to do the whole processing of the CSV file asynchronously if we don't want to kill our servers with potentially large files. These would be the steps:

* Upload file and enqueue a background job
* Said background job takes the uploaded temporal file as input and creates a batch of jobs using https://github.com/breamware/sidekiq-batch, as described on https://github.com/mperham/sidekiq/wiki/Batches (Note Sidekiq Batches is a Pro feature but that gem claims to be a drop-in replacement)
* Each job in the batch takes one row from the CSV file and does the parsing and processing (register, authorize or revoke)
* We send an email notification to the admin when done

Given `UserProcessor` currently process all entries serially, we'll need to refactor things a bit to allow for parallelization, however, I plan to make it backward compatible; to add parallel processing without removing the existing behavior.

This solution should let us spread the load and horizontally scale the system by adding more workers if needed.

**Comment by microstudi** [comment]
Looks nice, but does this mean that we will be force to use sidekiq? what about other queue processors, delayed_job, etc.

**Comment by sauloperez** [comment]
yes, I was mistakenly assuming Decidim's community mostly uses Sidekiq. Either 1) we ask for that or 2) check if there's an equivalent for Delayed Job and Resque, or 3) implement that strategy ourselves. I think we can go with 1 which will solve our most pressing issue and then check 2).

It'd be great if we could stick to 1) and have someone provide support for other backends as a contribution, if they need it.

**Comment by microstudi** [comment]
@sauloperez I don't think we can ask the Decidim community to use one specific queue. There's lots of installations using delayed_job. And I know some using rescue.
We really shoud try to make it queue-indepent. Otherwise this gem, that has no dependencies or migrations, will have a strong dependency.

**Comment by sauloperez** [comment]
ok, I'll investigate a bit further.

**Comment by sauloperez** [comment]
Alright, so the alternatives I found are tightly coupled to Redis (they rely on its data types and features) or are simply too complex for our use case. I have explored how a solution using Active Job's callbacks would look like and I really believe https://github.com/coopdevs/decidim-coopcat/pull/121 is all we need.

This is how it looks in practice (Note log lines are not in order of execution).

```rb
irb(main):018:0> DummyBatch.perform_later
Enqueued DummyBatch (Job ID: da0a831f-4fe5-4e35-bc9e-54034bf9ca15) to Async(default)
=> #<DummyBatch:0x00005644d90a7e80 @arguments=[], @job_id="da0a831f-4fe5-4e35-bc9e-54034bf9ca15", @queue_name="default", @priority=nil, @executions=0, @provider_job_id="ccd3221c-0ebf-4d2c-9fae-6741123d3994">
irb(main):019:0> Performing DummyBatch (Job ID: da0a831f-4fe5-4e35-bc9e-54034bf9ca15) from Async(default)
  => Persisting batch
   (0.7ms)  BEGIN
  Batch Create (1.2ms)  INSERT INTO "batches" ("job_id", "created_at", "updated_at") VALUES ($1, $2, $3) RETURNING "id"  [["job_id", "da0a831f-4fe5-4e35-bc9e-54034bf9ca15"], ["created_at", "2021-03-24 17:05:47.366996"], ["updated_at", "2021-03-24 17:05:47.366996"]]
   (1.8ms)  COMMIT  => Persisting initial state of job 1e420def-16a5-47f4-b0be-8e08b8345211Performing DummyJob (Job ID: 1e420def-16a5-47f4-b0be-8e08b8345211) from Async(default) with arguments: 26
  => Persisting end state of job 1e420def-16a5-47f4-b0be-8e08b8345211
   (0.3ms)  BEGIN  BackgroundJob Create (0.6ms)  INSERT INTO "background_jobs" ("status", "job_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["status", "enqueued"], ["job_id", "1e420def-16a5-47f4-b0be-8e08b8345211"], ["created_at", "2021-03-24 17:05:47.395400"], ["updated_at", "2021-03-24 17:05:47.395400"]]
  BackgroundJob Update All (0.8ms)  UPDATE "background_jobs" SET "status" = 'success' WHERE "background_jobs"."job_id" = $1  [["job_id", "1e420def-16a5-47f4-b0be-8e08b8345211"]]   (0.5ms)  SELECT COUNT(*) FROM "background_jobs" WHERE "background_jobs"."status" = $1  [["status", "enqueued"]]   (1.5ms)  COMMIT
  => 0 jobs remaining
  CACHE  (0.0ms)  SELECT COUNT(*) FROM "background_jobs" WHERE "background_jobs"."status" = $1  [["status", "enqueued"]]Enqueued DummyJob (Job ID: 1e420def-16a5-47f4-b0be-8e08b8345211) to Async(default) with arguments: 26  => Persisting initial state of job b2dbcfdd-4f6b-4e91-8cd4-f0c696506ed7
   (0.3ms)  BEGINPerforming DummyJob (Job ID: b2dbcfdd-4f6b-4e91-8cd4-f0c696506ed7) from Async(default) with arguments: 26  => Persisting end state of job b2dbcfdd-4f6b-4e91-8cd4-f0c696506ed7
  BackgroundJob Create (0.5ms)  INSERT INTO "background_jobs" ("status", "job_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["status", "enqueued"], ["job_id", "b2dbcfdd-4f6b-4e91-8cd4-f0c696506ed7"], ["created_at", "2021-03-24 17:05:47.400349"], ["updated_at", "2021-03-24 17:05:47.400349"]]
   (1.1ms)  COMMIT
Enqueued DummyJob (Job ID: b2dbcfdd-4f6b-4e91-8cd4-f0c696506ed7) to Async(default) with arguments: 26
Performed DummyBatch (Job ID: da0a831f-4fe5-4e35-bc9e-54034bf9ca15) from Async(default) in 66.67ms
Enqueued DummyCallback (Job ID: 9514e9f4-81cb-4fcf-a707-d01931c92125) to Async(default) with arguments: 26
Performed DummyJob (Job ID: 1e420def-16a5-47f4-b0be-8e08b8345211) from Async(default) in 21.22ms
Performing DummyCallback (Job ID: 9514e9f4-81cb-4fcf-a707-d01931c92125) from Async(default) with arguments: 26
  => Batch successfully executed
  => Notifying user...
  Batch Exists (0.3ms)  SELECT  1 AS one FROM "batches" WHERE "batches"."id" = $1 AND "batches"."finished_at" IS NULL LIMIT $2  [["id", 26], ["LIMIT", 1]]
  => Statuses cleaned up
Performed DummyCallback (Job ID: 9514e9f4-81cb-4fcf-a707-d01931c92125) from Async(default) in 1.24ms
  BackgroundJob Update All (1.7ms)  UPDATE "background_jobs" SET "status" = 'success' WHERE "background_jobs"."job_id" = $1  [["job_id", "b2dbcfdd-4f6b-4e91-8cd4-f0c696506ed7"]]
   (0.4ms)  SELECT COUNT(*) FROM "background_jobs" WHERE "background_jobs"."status" = $1  [["status", "enqueued"]]
  => 1 jobs remaining
  CACHE  (0.0ms)  SELECT COUNT(*) FROM "background_jobs" WHERE "background_jobs"."status" = $1  [["status", "enqueued"]]
Performed DummyJob (Job ID: b2dbcfdd-4f6b-4e91-8cd4-f0c696506ed7) from Async(default) in 14.8ms
```

**Comment by microstudi** [comment]
I've made a quick glance, and looks good to me! go ahead with this options, thanks!

**Comment by sauloperez** [comment]
Lol that was fast! I was thinking, we might not even need to implement it so abstract. It's fine if we have a CSVImportBatch job and a CSVRowJob or similar. We better keep it as simple as possible.

**Comment by microstudi** [comment]
@sauloperez sure, we will have to add a migration though anyway. I guess is the price to pay

**Comment by sauloperez** [comment]
I'm moving this forward in https://github.com/coopdevs/decidim-verifications-direct_verifications/tree/async-csv-import and I'll open a PR when ready.

**Comment by sauloperez** [comment]
Closed by #28
